design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+
+`role_v1:/realm1/tenant1//iam/manager`
+`role_v1:/realm1/tenant1//iam/developer`
+`role_v1:/realm1/tenant1//devops/developer`
+`role_v1:/realm1/tenant1//devops/devops_role`
+
+Now these roles are assigned directly to `UserA` and `UserB` so that:
+
+- `UserA` has the `role_v1:/realm1/tenant1//iam/manager` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+- `UserB` has the `role_v1:/realm1/tenant1//iam/developer` and the `role_v1:/realm1/tenant1//devops/developer` roles.
+
+Also, we should be able to still support giving everyone in a group an inherited role, such as:
+
+- `Group:devops` has the `role_v1:/realm1/tenant1//devops/devops_role` role.
+
+#### Representation in the tokens
+
+So now that we have a namespaced role attached to a user, the representation becomes inherently hierarchical, a **possible representation** in the tokens could look like:
+
+- For UserA
+
+```json
+{
+    ....
+    "roles": {
+        "tenants": {
+            "tenant1": {
+                "groups": {
+                    "iam": ["manager"],
+                    "devops": ["developer", "devops_role"]
+                }
+            }
+        }
+    }
+    ....
+}
+
+```
+
+- For UserB
+
+```json
+{
+    ....
+    "roles": {
+        "tenants": {
+            "tenant1": {
+                "groups": {
+                    "iam": ["developer"],
+                    "devops": ["developer", "devops_role"]
+                }
+            }
+        }
+    }
+    ....
+}
+```
+
+This also provides a way to isolate certain roles to specific tenants.
+
+- `role_v1:/realm1/tenant1//iam/developer` would only be visible in `tenant1` while
+- `role_v1:/realm1/tenant2//iam/somethingelse` would only be visible in `tenant2`.
+
+Or just to specific groups within the realm:
+
+- `role_v1:/realm1///iam/developer` would only be assignable to users in Group `iam`.
+- `role_v1:/realm1///devops/devops` would only be assignable to users in Group `devops`.
+
+### Use Case 2 - Custom, resource server specific roles
+
+Another interesting use case is the creation of roles that are not directly restricted to Keycloak's entities such as Realms, Groups, Clients etc.
+
+Right now in Keycloak, clients are effectively a representation of a Resource Server (among other things). You can assign roles to clients and those roles would only be usable on those clients.
+
+As previously mentioned, a role would be namespaced to a certain client such as:
+
+-  `role_v1:/realm1//clientId//roleName`
+
+A different way to be more flexible on giving users the possibility to set their own resource-server specific roles would be to provide a free-form unique role identifier that can be parsed by protocol mappers or just shown in the tokens in a hierarchical way.
+
+For example, a free form role could look like:
+
+- `role_v1:/realm1/tenant1///frole_v1:/mycompany/resources/department-a-roles/developer`
```

Comment:
I'm struggling a bit to understand this syntax. What does '////' imply, and why is there 'frole_v1:/'

Reply:
I'll add some more explanation on what the the `frole` is. But it's essentially a custom namespace created by users, while keeping the fixed path that contains all the variables to identify a role.

The forward slashes are basically there to keep separation and make it possible to access a variable by its position in the path.

Reply:
I don't think we need/should have the fixed path part at all to be honest.